docs/swift/DesignDoc.md

```diff
@@ -137,65 +152,117 @@ lets users treat messages polymorphically. Any shared method implementations
-    nilled out).
-*   Reading the value of a property returns its value if it is set, or returns
-    its default value if it is not set. Reading a property never returns nil.
-
-The final point in the list above implies that the optional cannot be checked to
-determine if the field is set to a value other than its default: it will never
-be nil. Instead, we must provide `has*` methods for each field to allow the user
-to check this. These methods will be public in proto2. In proto3, these methods
-will be private (if generated at all), since the user can test the returned
-value against the zero value for that type.
+The user-facing API for a `ProtoMessage` will be as follows:
+
+```swift
+public protocol ProtoMessage {
+
+  var serializedSize: Int { get }
+
+  init()
+  init(data: NSData) throws
+
+  mutating func merge(fromData data: NSData) throws
+
+  func data() throws -> NSData
+  func write(toOutputStream stream: NSOutputStream) throws
+
+  // Public only as an implementation detail:
+
+  mutating func merge(fromCodedInputStream stream: CodedInputStream) throws
+  func write(toCodedOutputStream stream: CodedOutputStream) throws
+}
+```
+
+The methods listed at the end are public only as an implementation detail,
+because the generated code will be implemented in terms of them to read and
+write the proto wire format. Users should never need to call them directly or
+use the coded stream types. The remaining methods and initializers are
+implemented in a protocol extension in terms of those coded stream methods.
+
+Conformance to `Equatable`, `Hashable`, and `CustomDebugStringConvertible` are
+also desirable but not currently shown, to focus on a small API intended for
+mobile clients. The implementations and generation of those methods are fairly
+straightforward, but it should be noted that generating equality and hashability
+for very large messages that are not likely to use those features would add
+significant bloat to an application. (Mirrors could be investigated as a way of
+reducing this duplication using a visitor pattern, albeit with some performance
+trade-offs.)
+
+### Fields
+
+Fields in a message will be represented as non-optional properties. We explored
+using optionals (regular and implicitly unwrapped), but determined that they
+were unsuitable; protocol buffer users expect to receive default values, not
+`nil`, when retrieving fields, and this is more important in proto2 where fields
+can have explicit non-zero defaults. Implicitly unwrapped computed properties
+could mitigate the `clear` situation by treating assignment of `nil` as clearing
+the field, but we would still need to create a `has` method because the property
+getter would never return nil (it would always return the default value, and
+there would be no way to distinguish "not set" from "explicitly set to the
+default value"). In that sense, having the property be optional seems misleading
+to users.
```

Comment:
We could look at killing the parameterless initializer and replacing it with one that takes the required fields as arguments, but that probably makes usage a good deal more difficult, and it would prohibit the scenario where you create a "draft" message with some required fields left unset and use partial serialization to store it aside until you're ready to use full serialization to ship it off (and only at that point would you want the required check to kick in).

Just to make sure I'm clear on this, proto2 distinguishes between "a field is unset" and "a field has been explicitly set to a value that just happens to be equal to its default", right? In that case, we'd still be able to check the values at serialization time; if primitives are stored as optionals in the underlying CoW storage class, we can distinguish it that way and error out if the user tries to save an incomplete proto.